### ISSUE TITLE
Gunkolle and multiple instances

### ISSUE THREAD

**Issue by BobbyDuggart** [Dustio/Gunkolle#15]
What changes would I need to make in order to get this program to work with multiple instances of GFL? Right now the client focuses on whatever window is active, so whenever I click on the 2nd instance, Gunkolle also starts clicking on the 2nd instance. From what I remember Gunkolle uses window IDs right? Could I lock Gunkolle onto "NoxPlayer" and then lock another instance of Gunkolle onto "NoxPlayer(1)"? Thanks again for your continued support on this project.

**Comment by Dustio** [comment]
Multi-instances use the same ahk window ID. I think you would need to use two different emulators.

**Comment by BobbyDuggart** [comment]
hmm okay that could work but is also hella annoying lmao. Is there any way I could lock Gunkolle to a set of coordinates on the screen, since the window never moves?

**Comment by Dustio** [comment]
I've never tried using FindClick without setting a window, so idk if it would work. You can use no window for global (screen) coordinates though. This would all take a while to set up and change all coords.

**Comment by BobbyDuggart** [comment]
Hmm okay, could I use global coordinates and set an offset somehow, so, instead of changing all the values I'd only have to change the offset?

**Comment by Dustio** [comment]
Offset should be there as long as you're using ClickS, although it currently uses window activation. You have to remove the window focus and use screen coordinates instead of window (what is currently being used). I don't know how to run the script without a focused window; I've never tried or looked into it. Maybe start with Window.ahk and the WINID in the config.

**Comment by BobbyDuggart** [comment]
Okay, ill mess with it myself a little later. If you come up with anything let me know. Thanks for the help.

**Comment by BobbyDuggart** [comment]
thanks for the DJMax update!
Could you explain what this line does (line 15 in Window.ahk)
`FindClick(A_ScriptDir "\pics\WaitForHome", "rLDPlayer mc o30 Count1 n0 a1200,,,-600")`

For some context, I changed line 9 in config.ini to `WINID=ahk_pid 224`
Previously this was `WINID=ahk_class LDPlayerMainFrame`
This makes Gunkolle use Process ID instead of Window Class. Multiple instances of LDPlayer have different PIDs, so I'm able to distinguish one specific instance from another (PIDs change every time I restart the client, which is annoying, but for now it works).
However, this only works in Window.ahk.
At line 15 in Window.ahk this happens:
`FindClick(A_ScriptDir "\pics\WaitForHome", "rLDPlayer mc o30 Count1 n0 a1200,,,-600")`
I assume this line is what makes the transfer from Window.ahk to FindClick.ahk.
Window.ahk was easy enough to figure out, but I have no fucking idea how FindClick.ahk works lol.
If I don't fix FindClick.ahk, Gunkolle will locate the correct window based on PID, but will prioritize any active LDPlayer window regardless of PID ("active" meaning when I click on it). Basically Window.ahk is using PID, but FindClick.ahk is still using HWND.

There are some lines in FindClick.ahk that seem to do things I don't want, but ; noting them out does nothing.

Some examples:
Line 853:
`WinSet, %@1%, %@2%, ahk_id %ThisHWND%`

Line 1105 and 1136:
`Hotkey, IfWinActive, ahk_id %GuiHWND%`

Line 1359:
`Hotkey, IfWinExist, ahk_id %GuiHWND%`

Not asking you to solve my problem for me or anything, but you seem to have more insight about FindClick than I do. Any help would be appreciated.

**Comment by Dustio** [comment]
`FindClick(A_ScriptDir "\pics\WaitForHome", "rLDPlayer mc o30 Count1 n0 a1200,,,-600")`

This uses findclick to search for the image"WaitForHome" one time and it clicks the image zero times (n0). It also does this in the area of the window I believe (a1200,,,-600). I'm not sure how to set the area exactly. I looked into it a little before and tried to remove the area completely, but it didn't make a difference for me. Maybe if you can get it to search in an area from your screen and not the window it can work.

I don't know if FindClick is using hwnd atm since I never entered it in anywhere except for Mouse.ahk (RenderWindow1). Maybe it is just activating the last window that was used still.

Hope this can help.

**Comment by BobbyDuggart** [comment]
yoo I figured it out. ignore the whole PID thing, it's actually a lot more simple than that.
`FindClick(A_ScriptDir "\pics\WaitForHome", "rLDPlayer mc o30 Count1 n0 a1200,,,-600")`
the "r" option in FindClick is used to specify a window. In the same way that I switched from LDPlayer to Nox by changing rLDPlayer to rNox. I can tell FindClick what window to target by changing the letters after the "r".
For example. If instance 1 is named "WINDOW1". I replace every "rLDPlayer" with "rWINDOW1" in Gunkolle.ahk, Maps.ahk, Mouse.ahk, and Window.ahk.
Then I copy and paste the entire folder and, in the new folder, change every "rWINDOW1" to "rWINDOW2", since instance 2 is named "WINDOW2".
Thanks for the help!
And one last thing as well. I recently bought a 4k monitor (hence why I wanted to try multiple instances). Gunkolle works perfectly on this monitor except for one thing. When at the home screen, it sometimes clicks the top bar where your commander level and resources are. It happens after finishing a 0-2 run. I think it clicks to verify that your at the home screen, but the randomization is too much and sometimes it clicks the resources, or "App Center" button on Nox. Do you know what I could do to fix this?

**Comment by Dustio** [comment]
Oh nice, that's a good easy fix then. 

The other thing is probably an issue with the GoHome function in Maps.ahk. Try to move the y coordinate on line 326, it might be clicking too high for you on Nox.

**Comment by BobbyDuggart** [comment]
yup that fixed it thanks!
Another thing while I'm here. I changed the MinWait and MaxWait while messing around with the client. It ended up breaking everything so I changed them back. However, the wait timers still didn't change back even after I checked everything in the files and config (and restarted my computer). Is there some sort of cache or something that's causing the MinRandomWait and MaxRandomWait variables not to change back?
